### PR TITLE
WIP: Named tracer prototype

### PR DESCRIPTION
**Pull request by rghetia**
This is a prototype for named tracer. What is in here?
- Global Provider Register (instead of Tracer Registry). This may not be required.
- SDK provides NewProvider method to create instance of Provider. Multiple provider instance can be created. This can help to parallelized testing.
   - It accepts optional parameter to associate exporters and configuration (sampler and other parameters) during instance creation.
   - exporters and configuration can also be applied at a later time.
- Some Tests are updated to created its own provider. Remove reference to global tracer.
  - few tests still uses Global Tracer.

### CODE REVIEW

File: sdk/trace/benchmark_test.go [hunk 1]
```diff
@@ -144,6 +145,7 @@ func BenchmarkSpanWithAttributes_all_2x(b *testing.B) {
 }
 
 func BenchmarkTraceID_DotString(b *testing.B) {
+	getTracer(b, "Benchmark traceID to string")
```
Comment (jmacd) [inline sdk/trace/benchmark_test.go]: Is this needed?

File: sdk/trace/benchmark_test.go [hunk 2]
```diff
@@ -157,6 +159,7 @@ func BenchmarkTraceID_DotString(b *testing.B) {
 }
 
 func BenchmarkSpanID_DotString(b *testing.B) {
+	getTracer(b, "Benchmark spanID to string")
```
Comment (jmacd) [inline sdk/trace/benchmark_test.go]: (same)

File: api/trace/api.go [hunk 3]
```diff
@@ -24,6 +24,13 @@ import (
 	"go.opentelemetry.io/api/distributedcontext"
 )
 
+type Provider interface {
+	// GetTracer creates a named tracer that implements Tracer interface.
+	// If name is an empty string then default name from the manager
+	// is used.
+	GetTracer(name string) Tracer
```
Comment (jmacd) [inline api/trace/api.go]: I wonder if this should just be `Get(name string) Tracer`.

**Comment by jmacd** [comment]
My concern about naming is that 

```
   tracer := trace.GlobalProvider().GetTracer(name) 
```

is fairly long. Renaming `GetTracer()` to `Get()` may help. In #216 I've proposed that globals move out of the core packages, so then I could imagine a package named `global` that would give us calls like:

```
  tracer := global.GetTracer(name)
```

the `Init()` method proposed in OTEP 0005 would take both a metric and a trace provider, but the caller doesn't need to write "provider" when accessing the globals.

File: example/http-stackdriver/client/client.go [hunk 4]
```diff
@@ -48,15 +45,20 @@ func initTracer() {
...
+		sdktrace.WithSyncer(exporter))
+	if err != nil {
+		log.Fatal(err)
+	}
+	trace.SetGlobalProvider(tp)
 }
 
 func main() {
 	initTracer()
+	tr := trace.GlobalProvider().GetTracer("stackdriver/example/server")
```
Comment (akvanhar) [inline example/http-stackdriver/client/client.go]: Should this be `/client`?

File: plugin/othttp/server_example_test.go [hunk 5]
```diff
@@ -44,22 +44,19 @@ func ExampleNewHandler() {
...
 	}
 
-	// Wrap stdout exporter with SimpleSpanProcessor and register the processor.
-	ssp := sdktrace.NewSimpleSpanProcessor(exporter)
-	sdktrace.RegisterSpanProcessor(ssp)
-
-	// For the example, use sdktrace.AlwaysSample sampler to sample all traces.
-	// In a production application, use sdktrace.ProbabilitySampler with a desired probability.
-	sdktrace.ApplyConfig(sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()})
+	//import sdktrace "go.opentelemetry.io/sdk/trace"
```
Comment (akvanhar) [inline plugin/othttp/server_example_test.go]: This comment doesn't make sense anymore.

File: sdk/trace/batch_span_processor_test.go [hunk 6]
```diff
@@ -181,16 +178,16 @@ func createAndRegisterBatchSP(t *testing.T, option testOption, te *testBatchExpo
 	if ssp == nil {
 		t.Errorf("%s: Error creating new instance of BatchSpanProcessor, error: %v\n", option.name, err)
 	}
-	sdktrace.RegisterSpanProcessor(ssp)
+	//sdktrace.RegisterSpanProcessor(ssp)
```
Comment (akvanhar) [inline sdk/trace/batch_span_processor_test.go]: delete this comment

File: sdk/trace/provider.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+		for k, v := range old {
+			new[k] = v
+		}
+	}
+	new[s] = &sync.Once{}
+	p.spanProcessors.Store(new)
+}
+
+// UnregisterSpanProcessor removes from the list of SpanProcessors the SpanProcessor that was
+// registered with the given name.
```
Comment (akvanhar) [inline sdk/trace/provider.go]: This comment is quite stuttery.
Suggestion:
`UnregisterSpanProcessor removes the given SpanProcessor from the list of SpanProcessors`

File: sdk/trace/span_processor_test.go [hunk 8]
```diff
@@ -48,10 +42,13 @@ func (t *testSpanProcesor) Shutdown() {
 
 func TestRegisterSpanProcessort(t *testing.T) {
 	name := "Register span processor before span starts"
+	tp, _ := sdktrace.NewProvider(
+		sdktrace.WithConfig(sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}))
```
Comment (akvanhar) [inline sdk/trace/span_processor_test.go]: This line is repeated a lot.
suggestion:
Set `var testConfig = sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}` and use that in `WithConfig`

File: api/trace/noop_trace_manager.go [hunk 9]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2019, OpenTelemetry Authors
```
Comment (paivagustavo) [inline api/trace/noop_trace_manager.go]: Should this file be named `noop_trace_provider.go`?

**Comment by rghetia** [comment]
> My concern about naming is that
> 
> ```
>    tracer := trace.GlobalProvider().GetTracer(name) 
> ```
> 
> is fairly long. Renaming `GetTracer()` to `Get()` may help. In #216 I've proposed that globals move out of the core packages, so then I could imagine a package named `global` that would give us calls like:
> 
> ```
>   tracer := global.GetTracer(name)
> ```
> 
> the `Init()` method proposed in OTEP 0005 would take both a metric and a trace provider, but the caller doesn't need to write "provider" when accessing the globals.

I'll take care of this in another PR.

File: sdk/trace/benchmark_test.go (hunk 1 above)
Comment (jmacd) [reply sdk/trace/benchmark_test.go]: It just doesn't look like the tracer is used in this benchmark. No big deal.

Comment (rghetia) [reply sdk/trace/benchmark_test.go]: I was looking in the wrong place. Yes, it is not needed. I have removed it.

**Pull request merged**